### DISCUSSION TITLE
Bulk inserting performance

### DISCUSSION THREAD

**Discussion by paradis-A** [appy-one/acebase#70] [Show and tell]
OS: Windows 10 64bit
RAM: 8gb
STORAGE: 128gb SSD M.2
CPU: Intel(R) Core(TM) i7-8550U CPU @ 1.80GHz  1.99 GHz
---------------------------------First run---------------------------------  
  Insert Non-indexed
    √  (767ms)

  Insert Indexed
    √  (4727ms)

  Get count of child (non-indexed)
     Childs:  2000
    √  (345ms)

  Get count of child (indexed)
     Childs:  2000
    √  (39ms)

  Non-Indexed instance.ref Foreach performance
    √  (800ms)

  Indexed instance.ref Foreach performance
    √  (740ms)

  Non-Indexed Query Foreach performance
    √  (924ms)

  Indexed Query Foreach performance
    √  (676ms)

  8 passing (9s)
  
---------------------------------Second run--------------------------------- 
  Insert Non-indexed
    √  (13152ms)

  Insert Indexed
    √  (23780ms)

  Get count of child (non-indexed)
     Childs:  4000
    √  (667ms)

  Get count of child (indexed)
     Childs:  4000
    √  (87ms)

  Non-Indexed instance.ref Foreach performance
    √  (1434ms)

  Indexed instance.ref Foreach performance
    √  (1369ms)

  Non-Indexed Query Foreach performance
    √  (1644ms)

  Indexed Query Foreach performance
    √  (1261ms)

  8 passing (43s)

---------------------------------Thrid run--------------------------------- 
  Insert Non-indexed
    √  (15832ms)

  Insert Indexed
    √  (29385ms)

  Get count of child (non-indexed)
     Childs:  6000
    √  (888ms)

  Get count of child (indexed)
     Childs:  6000
    √  (100ms)

  Non-Indexed instance.ref Foreach performance
    √  (2238ms)

  Indexed instance.ref Foreach performance
    √  (2083ms)

  Non-Indexed Query Foreach performance
    √  (2672ms)

  Indexed Query Foreach performance
    √  (2007ms)

  8 passing (55s)

Does update works like transaction where it loads all props then does the update stuff?
 
 See [Test Code](https://github.com/paradis-A/lendb-server/blob/main/test/acebaseIndex.spec.js)

**Comment by appy-one** [comment]
I've been working on improving the performance past week, almost ready to commit - one issue still needs to be ironed out. I wonder what your stats will look after those updates!

**Reply by paradis-A** [reply]
Just wow! I wish I can donate. Just am still seeking fortune out of this database hopefully more people your work when i'm ready to sell my products (sitting on top of this -Acebase-) in the market. Goodluck us!

**Comment by appy-one** [comment]
I've updated to version 1.15.0, you should notice significant performance improvements on these tests. Can you update and run them again? Would love to see new results 😄

**Reply by paradis-A** [reply]
you rock! it actually does


-----------first run------------
  Insert Non-indexed
    √  (770ms)

  Insert Indexed
[perf] tree rebuild took 5ms
[perf] getLeafStartKeys: 1ms
[perf] getEntries: 1ms
[perf] tree.create: 4ms
[perf] tree rebuild took 5ms
[perf] getLeafStartKeys: 1ms
[perf] getEntries: 0ms
[perf] tree.create: 5ms
[perf] tree rebuild took 12ms
[perf] getLeafStartKeys: 0ms
[perf] getEntries: 3ms
[perf] tree.create: 12ms
[perf] tree rebuild took 13ms
[perf] getLeafStartKeys: 0ms
[perf] getEntries: 4ms
[perf] tree.create: 13ms
[perf] tree rebuild took 22ms
[perf] getLeafStartKeys: 0ms
[perf] getEntries: 6ms
[perf] tree.create: 22ms
    √  (3783ms)

  Get count of child (non-indexed)
     Childs:  2000
    √  (331ms)

  Get count of child (indexed)
     Childs:  2000
    √

  Non-Indexed instance.ref Foreach performance
    √  (737ms)

  Indexed instance.ref Foreach performance
    √  (776ms)

  Non-Indexed Query Foreach performance
    √  (818ms)

  Indexed Query Foreach performance
    √  (657ms)


  8 passing (8s)

---------second run-----------
  Insert Non-indexed
    √  (936ms)

  Insert Indexed
[perf] tree rebuild took 32ms
[perf] getLeafStartKeys: 0ms
[perf] getEntries: 7ms
[perf] tree.create: 32ms
[perf] tree rebuild took 46ms
[perf] getLeafStartKeys: 0ms
[perf] getEntries: 10ms
[perf] tree.create: 46ms
    √  (3908ms)

  Get count of child (non-indexed)
     Childs:  3994
    √  (591ms)

  Get count of child (indexed)
     Childs:  4000
    √  (71ms)

  Non-Indexed instance.ref Foreach performance
    √  (1383ms)

  Indexed instance.ref Foreach performance
    √  (1428ms)

  Non-Indexed Query Foreach performance
    √  (1763ms)

  Indexed Query Foreach performance
    √  (1327ms)


  8 passing (11s)

---------third run-------------
 Insert Non-indexed
    √  (1021ms)

  Insert Indexed
[perf] tree rebuild took 97ms
[perf] getLeafStartKeys: 1ms
[perf] getEntries: 17ms
[perf] tree.create: 97ms
    √  (4600ms)

  Get count of child (non-indexed)
     Childs:  5990
    √  (871ms)

  Get count of child (indexed)
     Childs:  6000
    √  (98ms)

  Non-Indexed instance.ref Foreach performance
    √  (2072ms)

  Indexed instance.ref Foreach performance
    √  (2123ms)

  Non-Indexed Query Foreach performance
    √  (2649ms)

  Indexed Query Foreach performance
    √  (1985ms)


  8 passing (15s)

**Comment by paradis-A** [comment]
Excluding the query perfs. The bulk insert with or without index improved. Unlike before that it was like proportional based on the stored collection + inserts. Repetition of Bulk inserting now with the same amount (2000 in my case) improved.

Bulk Insert Before * 3 reps
1st Run
- Non-indexed (767ms)
- Indexed (4727ms)
2nd Run
- Non-indexed (13152ms)
- Indexed (23780ms)
3rd Run
- Non-indexed (15832ms)
- Indexed (29385ms)

Bulk Inser @v1.15.0 * 3 reps
1st Run
- Non-indexed (770ms)
- Indexed (3783ms)
2nd Run
- Non-indexed (936ms)
- Indexed (3908ms)
3rd Run
- Non-indexed (1021ms)
- Indexed (4600ms)
